crates/db/types/src/mmr_helpers.rs

```diff
@@ -224,10 +243,330 @@ pub fn find_peak_for_pos(pos: u64, mmr_size: u64) -> DbResult<u64> {
+    pub new_metadata: MmrMetadata,
+}
+
+/// Pure MMR algorithm logic separated from storage
+///
+/// This struct provides the core MMR algorithms (append, pop) as pure functions
+/// that compute what changes need to be made without actually making them.
+/// Storage implementations apply these changes within their own transaction context.
+#[derive(Debug)]
+pub struct MmrAlgorithm;
+
+impl MmrAlgorithm {
+    /// Compute the result of appending a leaf to the MMR
+    ///
+    /// # Arguments
+    ///
+    /// * `hash` - The hash value to append as a new leaf
+    /// * `metadata` - Current MMR metadata (num_leaves, mmr_size, peak_roots)
+    /// * `get_node` - Closure to read existing nodes (for sibling lookups during merge)
+    ///
+    /// # Returns
+    ///
+    /// `AppendLeafResult` containing:
+    /// - The leaf index of the newly appended leaf
+    /// - All nodes to write (leaf + internal nodes from merging)
+    /// - Updated metadata
+    ///
+    /// # Storage Independence
+    ///
+    /// This function doesn't know about transactions or storage backends.
+    /// The caller provides a closure to read nodes and applies the result
+    /// within their own transaction mechanism.
+    pub fn append_leaf<F, E>(
+        hash: [u8; 32],
+        metadata: &MmrMetadata,
+        get_node: F,
+    ) -> Result<AppendLeafResult, E>
+    where
+        F: Fn(u64) -> Result<[u8; 32], E>,
+        E: From<MmrError>,
+    {
+        let leaf_index = metadata.num_leaves;
+        let leaf_pos = leaf_index_to_pos(leaf_index);
+
+        let mut nodes_to_write = Vec::new();
+
+        // Store the leaf
+        nodes_to_write.push((leaf_pos, hash));
+
+        // Merge along the path to create internal nodes
+        let mut current_pos = leaf_pos;
+        let mut current_hash = hash;
+        let mut current_height = 0u8;
+
+        // Keep merging as long as we have a left sibling
+        loop {
+            let next_pos = current_pos + 1;
+            let next_height = pos_height_in_tree(next_pos);
+
+            // If next position is higher, current is a right sibling - we should merge
+            if next_height > current_height {
+                // Current is right sibling, get left sibling
+                let sibling_position = sibling_pos(current_pos, current_height);
+                let sibling_hash = get_node(sibling_position)?;
+
+                // Create parent hash (left sibling, right sibling)
+                let parent_hash = Sha256Hasher::hash_node(sibling_hash, current_hash);
+
+                // Store parent
+                nodes_to_write.push((next_pos, parent_hash));
+
+                // Move up to parent
+                current_pos = next_pos;
+                current_hash = parent_hash;
+                current_height = next_height;
+            } else {
+                // Current is a left sibling (will be merged when right sibling comes)
+                // or we've reached a peak - stop here
```

Comment:
even if we reached a peak, if a node isn't a right sibling then it is a potential left sibling (that will be merged when, and if, a right sibling comes), so you can safely remove the second case.

Reply:
I suggest we do this in the next optimization ticket for MMR. In this PR, I am not trying to modify the core algorithm itself as it was in the codebase.